Introduce a protocol version and a message for it

bertrand-marquis opened the pull request:
Introduce the VIRTIO_MSG_VERSION and define a protocol to negociate a protocol version between the driver and the device side.

Current version is set to 0.1.

Bump all message number so that the Version message is numbered 0x2 (0x1 being the Error message).

# Thanks for proposing a change to the virtio-msg specification

This is not the same as proposing a change to the VirtIO spec. This
repository is intended for preparing the virtio-msg transport
specification before it's submission to the VIRTIO TC.

If your intention was to suggestion to change to the upstream please
propose the change as a patch to the [mailing list](https://www.oasis-open.org/committees/tc_home.php?wg_abbrev=virtio#feedback).

bertrand-marquis committed:
Introduce a protocol version and a message for it

Introduce the VIRTIO_MSG_VERSION and define a protocol to negociate a
protocol version between the driver and the device side.

Current version is set to 0.1.

Bump all message number so that the Version message is numbered 0x2 (0x1
being the Error message).

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Make the messge size variable

Modify the message protocol to have a variable message size and allow
implementations that could tranfer bigger messages to benefit of all the
size available:
- add a maximum message size in the version message: to inform both
  sides of what the other can support and limit the size used to what
  can be transfered.
- add a message payload size field in the header
- modify get/set feature and config to have a variable size and be able
  to transfer as much data as possible (rework the offset to support any
  offset value and adapt to the different possible sizes)
- rework bus device discovery message to be able to use a bigger bitmap

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Introduce the configuration generation count

Introduce a configuration generation count and describe it in a specific
chapter.

Modify the following messages to use it:
- get_config: provide the value in the answer
- set_config: add in request, provide in answer, error if not right in
  request.
- get_config_gen: remove message
- event_config: provide the value

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Give hints in bus message of what can be done

Mention that a bus implementation can provide a solution for memory
sharing or could rely on MSI-x for notification instead of actually
sending an event.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

arnopo commented on transport-msg.tex:
```diff
@@ -137,6 +139,58 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+A Virtio Message Protocol Version is encoded in 32 bits with:
+\begin{itemize}
+    \item Bits[31-16]: Major Version (Current is 0).
+    \item Bits[15-0]: Minor Version (Current is 1).
```
i wonder if we need a major and minor version.
- compliance with other transport layer: virtio MMIO uses a simple version
- What would be  the criteria for minor and major version evolution?
- can be complexe to manage in code versions with minor and Major,

bertrand-marquis replied:
I started from the proposal from Bill to have version 0.1 and 0.x until first release of the spec.
But I am open to have a version starting at 0 and increasing (which would mean it would be 1 right now)

arnopo commented on transport-msg.tex:
```diff
@@ -184,10 +188,10 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
 \hline
 Answer & 0 & 4 & Device Protocol version \\
-& 4 & 32 & Reserved (MBZ) \\
+& 4 & 4 & Maximum message size supported \\
```
it is quite confusing for me 

In  the message property description  you say:
```
A message has a minimum size of 8 bytes and any implementation must support
  to transmit messages of up to 64 bytes.
```
does it means that the  size supported is from "Maximum message size supported" size to 64 bytes ? or does it means that the implementation have to support from 64 bytes to "Maximum message size supported"

bertrand-marquis replied:
ah yes i struggled a bit here.
Here is my train of thought, i would need help to confirm and express:
- minimum message size is 8 bytes (header size + 0 payload)
- we would like any transport to support messages of at least 64bytes (meaning that they could support more but less will start to be challenging for config or feature messages)
- in practice we must have at least 44 bytes (header size + 36 for the set/get vqueue message)

So it might be easier to just say we need at least 44 bytes.
Now i need to make a pass on the size so that we always speak of header size + payload size.
What do you think ?

arnopo commented on transport-msg.tex:
```diff
@@ -203,24 +207,30 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
 Type & Offset & Size (bytes) & Content \\
 \hline \hline
-Request & 0 & 36 & Reserved (MBZ) \\
+Request & 0 & 0 & None \\
```
payload size is 0 here , so minimum payload size is 0 not 8, right ?

bertrand-marquis replied:
8 is with the header, so minimum payload is 0 but minimum message size is 8

arnopo commented on transport-msg.tex:
```diff
@@ -232,21 +242,25 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
-Request & 0 & 4 & Feature index \\
-& 4 & 32 & Reserved (MBZ) \\
+Request & 0 & 4 & Feature bit index \\
+& 4 & 4 & Number of features \\
```
Does it means here that I can ask 35 feature bits from bit number 3?

should we have feature bit index =  X*32 to simplify implementations

bertrand-marquis replied:
In my current proposal it is the case so that we can use as much space as available in the message.
Now i am open to suggestion if we think that we should only allow a multiple of 32 (or 8, or 16).

arnopo commented on transport-msg.tex:
```diff
@@ -604,11 +620,14 @@ \subsubsection{Message Definition}\label{sec:Virtio Transport Options / Virtio O
+A request at offset X gives the availability of device IDs starting at X with
+a number of IDs depending on the actual message size received:
+
+X to (X + Message Size)*8 - 1
```
What about using a field to define the number of devices ID to get , as done for the feature and the config (homogeneity of messages)

bertrand-marquis replied:
Would make sense yes, i can fix that.

arnopo commented on transport-msg.tex:
```diff
@@ -62,6 +62,27 @@ \subsection{Errors Definition}\label{sec::Virtio Transport Options / Virtio Over
+
+This system allows the driver to maintain a cache of the configuration space
+of a device and return values from it instead of asking the device when a
+configuration value is needed.
```
I'm trying to understand how it work  from virtio driver perspective. 
let's say that the device update the configuration. Is there a specific mechanism  to notify the virtio driver from the virtio transport that the config has changed?

bertrand-marquis replied:
When the device does an update in the configuration, it has to send an EVENT_CONFIG message to notify the driver that something changed in the config

bertrand-marquis committed:
Rework version message

Change the version from major.minor to an unsigned integer that we will
increase each time the protocol is modified and define current version
as "1".
Remove the version for the device info message as we have a specific
message for it.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Update the message size definition

Update the message size definition to be clearer:
- minimum message size is 8 bytes (header and empty payload)
- minimum message size that a transport must support is 44 bytes
  (virtqueue set message)

Add a sentence to explain how size negocation works.

Modify get/set feature to use 32 bit blocks instead of allowing to
request any bit offest and size.

Modify bus message to discover devices to also include a number of
device IDs requested in the message (instead of relying on the message
size) to be more coherent with the other messages.

Add a mention in what a bus can do for restricting size and protocol
version that are negociated to restrict them to what it can handle.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Add a mention of the event message for the config cache

Add a sentence to signal that any change of the configuration not
requested by the driver is signaled to the driver by the device through
the event message.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

arnopo commented on transport-msg.tex:
```diff
@@ -198,6 +199,13 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+message it can receive. Each side must save this information and ensure not
+to send messages bigger than what the other side can handle. It is generally
+expected that both side will not send message bigger than the biggest size
+supported by both sides.
```
Is the last sentence true?
if the driver supports a maximum  of  44 bytes message max size and the device supports 64 bytes,
it would mean that the device can send a 64-bytes message to the driver.

if we use static buffers for message exchanges this should be not possible.

It seems to me that the sentence should be: 
It is generally expected that both sides will not send message bigger than the "smallest" size
supported by both sides.

bertrand-marquis replied:
ack

arnopo commented on transport-msg.tex:
```diff
@@ -544,6 +552,7 @@ \subsection{Message Bus}\label{sec:Virtio Transport Options / Virtio Over Messag
 \item One Message Bus has one or more devices on the backend side handled by
       drivers on the frontend side.
 \item An endpoint can be connected to several Message Buses.
+\item Any Message bus must support to transmit messages of at least 44 bytes.
```
"to transmit" or "transmitting"

bertrand-marquis replied:
ack

arnopo commented on transport-msg.tex:
```diff
@@ -565,6 +574,14 @@ \subsection{Message Bus}\label{sec:Virtio Transport Options / Virtio Over Messag
+        claims to support to reduce it to the maximum size it can support
+        and do the same with size claimed by the device. If a bus only support
+        specific version of the protocol it can also restrict what is claimed
+        by a driver or a device by modifying the message protocol version part.
```
That seems to me very complex to handle.
For me only the driver message should be intercepted by:
- the driver bus before sending the message
- the device bus if the version/size not supported ( interceprt the received message)
- the device if the version/size is not supported (at this point the driver and the device buses support the version)

bertrand-marquis replied:
I agree and i am wondering a bit on this:
- Should it matter for the bus of what protocol is used on top ? the bus only transmits virtio messages and should not care of what is inside them as long as the header does not change
- The maximum message size supported by the bus should be communicated to the transport by the bus on init through some impdef way (in linux possibly by some kind of field on the bus device)
In general i think we should prevent as much as possible to require the bus to intercept and modify messages.
The event case is specific and in fact means the bus is generating messages for the transport and i think that one is ok.
What do you think @arnopo ?

arnopo replied:
Yes I share your wonderrings


> * Should it matter for the bus of what protocol is used on top ? the bus only transmits virtio messages and should not care of what is inside them as long as the header does not change

This looks like we are facing the problem of the chicken and the egg if the bus uses the virtio-msg protocol to transmit its own messages.

let's try to dig into version needs.

- Down to top dependencies:

  - the bus uses the virtio-msg protocol to negotiate the version.

- Top to done dependencies:

  - the virtio devices/drivers (console, net, i2c, ballon, ...) should respect a kind of service version.
    updating this version should be transparent for the virtio-msg protocol but could also lead to the enhance/update of the virtio messages to support a new feature. 
  - the virtio-msg bus should have its own versioning to support an update of the shared resources (for instance  memory structuring).

Does it make sense that the bus fixes the virtio-msg version used on init? It would determine the version in a static way or using its own protocol to negotiate the version with the other side. 

> * The maximum message size supported by the bus should be communicated to the transport by the bus on init through some impdef way (in linux possibly by some kind of field on the bus device)

Yes, that would make sense, the bus could initialize the transport provided the version and max sizes supported. 

>   In general i think we should prevent as much as possible to require the bus to intercept and modify messages.
>   The event case is specific and in fact means the bus is generating messages for the transport and i think that one is ok.

Agree

bertrand-marquis replied:
Currently in the FF-A Virtio Message Bus specification, i create a a protocol version that is used by the bus itself for its own need (I am on version 1 and i plan to add an other transfer method with version 2 using a pre shared memory area).
The bus can use standard bus messages defined by the Virtio Message spec and that would be the only dependency somehow with the version of virtio message.
But a bus can on its own manage with its own version which standard bus messages are used or not (at the end it is sending and using those messages, not the transport).
So i think we could end up with a Virtio Message version which only impacts the transport but the bus does not care of and a bus implementation can define its own version for its own need.
The only caveats is to make sure that:
- the header does not change
- the bus does not need to understand or intercept any transferred virtio message
- the maximum message size is an information that the bus must give to the transport (in FF-A case it will be a static 96 value for now but a bus implementation could do its own negociation before informing the transport)
What do you think ?

arnopo replied:
> Currently in the FF-A Virtio Message Bus specification, i create a a protocol version that is used by the bus itself for its own need (I am on version 1 and i plan to add an other transfer method with version 2 using a pre shared memory area).

you can have a look to the memory structuring we discussed with @wmamills 
an implementation is available here:
https://github.com/arnopo/open-amp/blob/virtio-msg/lib/utils/amp-static-queue.c
https://github.com/arnopo/open-amp/blob/virtio-msg/lib/include/openamp/amp_static_queue.h

>The bus can use standard bus messages defined by the Virtio Message spec and that would be the only dependency somehow with the version of virtio message. But a bus can on its own manage with its own version which standard bus messages are used or not (at the end it is sending and using those messages, not the transport). So i think we could end up with a Virtio Message version which only impacts the transport but the bus does not care of and a bus implementation can define its own version for its own need. The only caveats is to make sure that:
>
> * the header does not change

This is my main concern.  should we have a specific VIRTIO_MSG_TYPE  for the version, with a specific immutable  message?

> * the bus does not need to understand or intercept any transferred virtio message

It should not be an issue if the virtio-msg supported version is fixed by the bus, right?

> * the maximum message size is an information that the bus must give to the transport (in FF-A case it will be a static 96 value for now but a bus implementation could do its own negociation before informing the transport)

Definitely agree with that. Should it be the same for the virtio-msg version supported?

bertrand-marquis replied:
> > Currently in the FF-A Virtio Message Bus specification, i create a a protocol version that is used by the bus itself for its own need (I am on version 1 and i plan to add an other transfer method with version 2 using a pre shared memory area).
> 
> you can have a look to the memory structuring we discussed with @wmamills an implementation is available here: https://github.com/arnopo/open-amp/blob/virtio-msg/lib/utils/amp-static-queue.c https://github.com/arnopo/open-amp/blob/virtio-msg/lib/include/openamp/amp_static_queue.h
> 
> > The bus can use standard bus messages defined by the Virtio Message spec and that would be the only dependency somehow with the version of virtio message. But a bus can on its own manage with its own version which standard bus messages are used or not (at the end it is sending and using those messages, not the transport). So i think we could end up with a Virtio Message version which only impacts the transport but the bus does not care of and a bus implementation can define its own version for its own need. The only caveats is to make sure that:
> > 
> > * the header does not change
> 
> This is my main concern. should we have a specific VIRTIO_MSG_TYPE for the version, with a specific immutable message?
> 

That could be a solution. But how to define this one so that it does not clash with the current header ?

> > * the bus does not need to understand or intercept any transferred virtio message
> 
> It should not be an issue if the virtio-msg supported version is fixed by the bus, right?

I just have one concern with that: After last meeting it was decided that the virtio msg version should be a per Device message (to be compliant with the MMIO transport). Having the protocol version enforced by the bus would go against that principle and would make the need for a per device version message a bit pointless.

> 
> > * the maximum message size is an information that the bus must give to the transport (in FF-A case it will be a static 96 value for now but a bus implementation could do its own negociation before informing the transport)
> 
> Definitely agree with that. Should it be the same for the virtio-msg version supported?

I would say yes but we definitely look at the per device protocol version angle.

Overall i start to think that we are making all this to complex and that saying that the bus is defining the protocol version and maximum message size and remove the version message from the virtio-msg messages should be possible.

arnopo replied:
> > > Currently in the FF-A Virtio Message Bus specification, i create a a protocol version that is used by the bus itself for its own need (I am on version 1 and i plan to add an other transfer method with version 2 using a pre shared memory area).
> > 
> > 
> > you can have a look to the memory structuring we discussed with @wmamills an implementation is available here: https://github.com/arnopo/open-amp/blob/virtio-msg/lib/utils/amp-static-queue.c https://github.com/arnopo/open-amp/blob/virtio-msg/lib/include/openamp/amp_static_queue.h
> > > The bus can use standard bus messages defined by the Virtio Message spec and that would be the only dependency somehow with the version of virtio message. But a bus can on its own manage with its own version which standard bus messages are used or not (at the end it is sending and using those messages, not the transport). So i think we could end up with a Virtio Message version which only impacts the transport but the bus does not care of and a bus implementation can define its own version for its own need. The only caveats is to make sure that:
> > > 
> > > * the header does not change
> > 
> > 
> > This is my main concern. should we have a specific VIRTIO_MSG_TYPE for the version, with a specific immutable message?
> 
> That could be a solution. But how to define this one so that it does not clash with the current header ?

The only way I can see is to consider the first byte of the message and based on the value determine the header.

But not convince by this solution. 
 
> 
> > > * the bus does not need to understand or intercept any transferred virtio message
> > 
> > 
> > It should not be an issue if the virtio-msg supported version is fixed by the bus, right?
> 
> I just have one concern with that: After last meeting it was decided that the virtio msg version should be a per Device message (to be compliant with the MMIO transport). Having the protocol version enforced by the bus would go against that principle and would make the need for a per device version message a bit pointless.

Yes, I remember the discussion. I had a look at the MMIO version. From my understanding, the MMIO version is not linked to the virtio devices but to the virtio MMIO transport. The device should expose the virtio MMIO registers with the version register initialized.

We could have the same approach, meaning the virtio device bus fixes the version.

More than that, if we multiplex devices on one bus, it seems simpler to manage only one version per bus instead of one version per device.

> 
> > > * the maximum message size is an information that the bus must give to the transport (in FF-A case it will be a static 96 value for now but a bus implementation could do its own negociation before informing the transport)
> > 
> > 
> > Definitely agree with that. Should it be the same for the virtio-msg version supported?
> 
> I would say yes but we definitely look at the per device protocol version angle.
> 
> Overall i start to think that we are making all this to complex and that saying that the bus is defining the protocol version and maximum message size and remove the version message from the virtio-msg messages should be possible.

I think we are sharing the same conclusion. :smiley:
As you proposed, the bus determines the version and the max size and initializes the virtio-message protocol with these configurations.
If that does not match with the virtio-message capabilities, it just means that you have an issue in your system. But we would reach the same state with some version messages.

We can perhaps have a new round of discussion on this topic in the Thursday meeting to conclude...

bertrand-marquis replied:
Definitely agree with you on that.
@wmamills, @stsquad Could we have your thoughts on this ?
The thursday meeting is canceled this week as Bill is not there i think and it would be good to make some progress on the subject.
If you agree on this i can make a proposal going this way.

wmamills commented:
I am closing this PR as all the commits here are also included in PR#16.
The discussion here is good and can still be seem by looking at closed PRs
